### PR TITLE
fix Issue 20053 - add mixin types

### PR DESCRIPTION
**Pull request by WalterBright**
Currently D has these mixins:

1. declaration
2. statement
3. expression

But lacks types. This adds mixin types. I.e. adding:
```
 MixinType:
    mixin ( ArgumentList)
```
to https://dlang.org/spec/declaration.html#BasicType

### CODE REVIEW

**Comment by WalterBright** [comment]
@TurkeyMan FYI

**Comment by thewilsonator** [comment]
home/circleci/dmd/src/dmd/typesem.d(1893): Error: template instance `Parser!ASTCodegen` template `Parser` is not defined

**Comment by TurkeyMan** [comment]
Thank you so much. This is really good.

**Comment by WalterBright** [comment]
Yah, it was a rather glaring gap in the mixin capability.

**Comment by WalterBright** [comment]
I'll try to get it through the test suite tomorrow.

**Comment by Geod24** [comment]
Please add the following to a fail_compilation test:
```d
struct Foo {
    alias T = mixin("T2");
}
alias T1 = mixin("Foo.T");
alias T2 = mixin("T1");
void func (T2 p) {}
```

Or any similar test with circular reference

**Comment by TurkeyMan** [comment]
@Geod24 Nice thinking :)

**Comment by WalterBright** [comment]
Spec: https://github.com/dlang/dlang.org/pull/2684

File: src/dmd/astbase.d [hunk 1]
```diff
@@ -4243,6 +4246,41 @@ struct ASTBase
...
+                    a = new Expressions(exps.dim);
+                    foreach (i, e; *exps)
+                    {
+                        (*a)[i] = e ? e.syntaxCopy() : null;
+                    }
+                }
+                return a;
+            }
+
+            return new TypeMixin(arraySyntaxCopy(exps));
```
Comment (andralex) [inline src/dmd/astbase.d]: I'm confused as to why this idiom was used here. It's just a more complicated way of saying... this:

```d
override Type syntaxCopy()
{
    Expressions* a = null;
    if (exps)
    {
        a = new Expressions(exps.dim);
        foreach (i, e; *exps)
        {
            (*a)[i] = e ? e.syntaxCopy() : null;
        }
    }
    return new TypeMixin(a);
}
```

File: src/dmd/parse.d [hunk 2]
```diff
@@ -1950,6 +1938,18 @@ final class Parser(AST) : Lexer
...
     }
 
+    /***************************************
+     * Parse a Type or an Expression
+     * Returns:
+     *  RootObject representing the AST
+     */
+    RootObject parseTypeOrAssignExp()
+    {
+        return (isDeclaration(&token, NeedDeclaratorId.no, TOK.reserved, null))
```
Comment (andralex) [inline src/dmd/parse.d]: superfluous parens

File: src/dmd/astbase.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/astbase.d]: Because `arraySyntaxCopy` is used all over dmd, but is not available in astbase.d

**Review by Geod24 — CHANGES_REQUESTED** [review]
Please also add a test for multiple arguments to `mixin` (e.g. `mixin("a", ".", "b");`)

**Review by atilaneves — APPROVED** [review]

**Comment by Basile-z** [comment]
This has been merged without

- changelog entry
- h file counterpart
- dlang.org PR for this.

Same standards for everyone please.

**Comment by WalterBright** [comment]
> h file counterpart

Don't need it.

> dlang.org PR for this.

https://github.com/dlang/dmd/pull/10215#issuecomment-514931648

**Comment by WalterBright** [comment]
> Please also add a test for multiple arguments to mixin (e.g. mixin("a", ".", "b");)

Not necessary due to common handling with other mixins. It's only necessary to test what is different.

**Comment by Basile-z** [comment]
changelog PR and we're good. Something inside [this folder](https://github.com/dlang/dmd/tree/master/changelog)

**Comment by Geod24** [comment]
> Not necessary due to common handling with other mixins. It's only necessary to test what is different.

[This line](https://github.com/dlang/dmd/pull/10215/files#diff-cd066d37445cac534313c0137c2d4bbeR3720) is not shared, and what I wanted to have covered by tests.

**Comment by WalterBright** [comment]
That line looks green to me, and `parseArguments()` is called in 13 other places in `parse.d`.

**Pull request merged**